### PR TITLE
Json pretty

### PR DESCRIPTION
**Pull request by guygastineau**
This is a migration of #98 for merging into `feature` instead of `master`.

### CODE REVIEW

File: code/json.sls [hunk 1]
```diff
@@ -30,172 +30,232 @@
...
-		(pair? (car x))
-		(not (pair? (caar x))))
-	   (write-ht x p))
-	  ((list? x) (write-array x p))
-	  (else (error "Invalid JSON object in json-write" x))))
+        (cond
+         ((or (string? x)
+              (number? x)) (write x p))
+         ((boolean? x) (display (if x "true" "false") p))
+         ((symbol? x)
```
Comment (guygastineau) [inline code/json.sls]: I think this is much nicer on the eyes than dropping the port argument to the next line

File: config.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
     {
       "slug": "hello-world",
       "uuid": "f2936612-edc4-44f9-a403-c60f7c6486e4",
-      "core": true,
+      "core": false,
```
Comment (guygastineau) [inline config.json]: Oh, I need to fix that.  I had to do a lot of manual merge conflict fixes while rebasing this.

File: code/track.ss [hunk 3]
```diff
@@ -9,38 +9,8 @@
...
-  (map (lambda (x)
-         (if (not (eq? (car x) 'exercises))
-             x
-             `(exercises . ,(remp (lambda (exercise)
-                                    (memq 'wip (map car exercise)))
-                                  (cdr x)))))
-       track-config))
-
-;; Check problem's entry in config for uuid and existence.
-(define (check-config-for problem)
```
Comment (guygastineau) [inline code/track.ss]: I didn't mean to remove this.  Let me see how to get it back.

File: code/track.ss [hunk 4]
```diff
@@ -9,19 +9,9 @@
       (delete-file config.json))
     (with-output-to-file config.json
       (lambda ()
-	(json-write (process-config))))))
-
-;; preprocess nodes in config. currently removes exercises marked wip
-(define (process-config)
```
Comment (guygastineau) [inline code/track.ss]: This has been moved to `outils.ss`

File: config.json (hunk 2 above)
Comment (guygastineau) [reply config.json]: Fixed

**Comment by guygastineau** [comment]
Most of the diff is still formatting fixes, and some cleaning up of trailing whitespace.  I am just super anal about trailing whitespace.

Other than that I am feeling pretty good about the patch on `json.sls` and the helpers I put in `outils.ss`, so let me know if this looks good to you.

Everything is working great. In fact, during the multistage merge conflict resolution, once I knew that I had the `config.ss` in line with yours I just used `make config.json` to fix the conflicts between our `config.json` files, so things are in good shape.

File: code/json.sls [hunk 5]
```diff
@@ -30,172 +30,232 @@
...
+      ;; control whether or not to pretty print
+      (define pretty? #f)
+      ;; control variables for pretty printing
+      (define indent-level 0)
+      (define tabstop-size 2)
+
+      ;; Print the indentation level
+      (define (display-level p)
+        (display (make-string (* indent-level tabstop-size) #\space) p))
+
```
Comment (jitwit) [inline code/json.sls]: This should probably be rewritten to 
```
(let ((pretty? #f) (indent-level 0) ...)
   ...)
```
rather than `defines`

File: code/track.ss (hunk 3 above)
Comment (jitwit) [reply code/track.ss]: Maybe to follow some of your advice it would be better to fix the trailing whitespace/formatting in a separate request? It makes it harder to see what the substantial additions are

File: code/json.sls [hunk 6]
```diff
@@ -30,172 +30,232 @@
...
+            (write-any obj (current-output-port)))
+           (else (error 'json-write "Invalid message" msg))))
+        ;; Pretty Print with custom tabsize defined in spaces
+        ((msg obj tabsize)
+         (case msg
+           ((pretty)
+            (set! pretty? #t)
+            (set! tabstop-size tabsize)
+            (write-any obj (current-output-port)))
+           (else (error 'json-write "invalid message" msg)))))))
```
Comment (jitwit) [inline code/json.sls]: I think perhaps the arguments should be reversed? It's seems more natural if the object to print appears before the flag to pretty print. What do you think?

File: code/json.sls (hunk 5 above)
Comment (jitwit) [reply code/json.sls]: Perhaps keep the helper functions as `define`s, however.

File: code/json.sls (hunk 6 above)
Comment (guygastineau) [reply code/json.sls]: Yeah, I am fine with that.

File: code/json.sls (hunk 5 above)
Comment (guygastineau) [reply code/json.sls]: You mean at the top level of the `json-write` function? Like in the as-of-yet ignored `()` after let?

I'll try it out.

Comment (guygastineau) [reply code/json.sls]: Cool.  Got up the alteration.  Let me know if it is what you meant.

File: code/json.sls (hunk 6 above)
Comment (guygastineau) [reply code/json.sls]: Reversed them and fixed the documenting comment at the top

**Comment by guygastineau** [comment]
I also regenerated the `config.json` with the refactoring to make sure there were no changes.  Git was satisfied that they are still the same file ;)

File: code/outils.ss [hunk 7]
```diff
@@ -1,37 +1,75 @@
...
+;; Top level helper for process-config
+(define (exercises->snake-case exercises)
+  (let ((handle (lambda (exercise)
+                  (map format-js-pair exercise))))
+    (map handle exercises)))
 
+;; Helpers for exercises->snake-case
+(define (kebab->snake str)
+  (let ((k->s (lambda (c) (if (char=? c #\-) #\_ c))))
+    (apply string (map k->s (string->list str)))))
```
Comment (jitwit) [inline code/outils.ss]: This should probably re-use `make-char-quotator` instead

File: code/track.ss (hunk 4 above)
Comment (jitwit) [reply code/track.ss]: I think it should stay in `track` since it deals specifically with the config. `outils` is intended for generally useful helpers. `process-config` is specific to selecting which problems should be used.

Comment (guygastineau) [reply code/track.ss]: Okay, that is fine with me.  I think at some point we should actually just move all of the config helpers into config-utils.ss.

Do you want me to go ahead and do that, or just move `preprocess-config` back to `track.ss`?

Comment (guygastineau) [reply code/track.ss]: Hmm, the helpers I made for `process-config` are currently only useful to the config stuff, but they are more general than the config stuff.

They seem like they are really formatting utilities to help with compatibility with the web site.

What do you think about storing them in a file called `compatibility-utils.ss`?

Then `process-config` would just go back into `track.ss`.

Comment (guygastineau) [reply code/track.ss]: I know I would have to modify `load.ss` then, so if you think that kind of refactoring should just happen in another PR, then I will go ahead with the simple suggestion you gave.

File: code/outils.ss [hunk 8]
```diff
@@ -1,37 +1,75 @@
...
 
+(define (symbol->snake-case-string symbol)
+  (kebab->snake (symbol->string symbol)))
+
+(define (format-js-pair pair)
+  (let ((snake-key (symbol->snake-case-string (car pair))))
+    (if (null? (cdr pair))
+        `(,snake-key)
+        ;; sort the topics list
+        (if (symbol=? 'topics (car pair))
```
Comment (jitwit) [inline code/outils.ss]: `eq?`

Comment (guygastineau) [reply code/outils.ss]: Would `eq?` be better here?  I figured the more specific `symbol=?` would be better.

Comment (jitwit) [reply code/outils.ss]: I think `symbol=?` is hardly ever used

File: code/track.ss (hunk 4 above)
Comment (jitwit) [reply code/track.ss]: I'm not sure I understand what `compatibility` means here. I think it does sound like these sorts of changes would be better from another pr. The json pretty-printing functionality shouldn't really be leaking from `json.sls`. Granted, it's specific to making exercism/configlet happy, so perhaps a separate file dedicated to it makes the most sense. It's not general enough to be in `json.sls`, and it's internals should be sealed somewhat.

File: code/outils.ss (hunk 8 above)
Comment (guygastineau) [reply code/outils.ss]: Okay.

Comment (jitwit) [reply code/outils.ss]: It's not. 0 times in `scmutils`. 1 time in the `srfi` collection. In `chezscheme` pretty much only used at the site of definition

Comment (jitwit) [reply code/outils.ss]: (according to grep)

File: code/track.ss (hunk 4 above)
Comment (guygastineau) [reply code/track.ss]: Yeah, so I wasn't meaning to have anything from `json.sls` leak into another file.

Basically the idea is just that the helpers that convert parts of the config to snake case (since some of them are general even though they only have one application now) could be put in a file that has a name concerning compatibility.  If everything were scheme all the time we wouldn't have to worry about it.

I'll go with your first suggestion for now ;)

File: code/outils.ss (hunk 8 above)
Comment (jitwit) [reply code/outils.ss]: [https://www.scheme.com/tspl4/objects.html#./objects:h2](https://www.scheme.com/tspl4/objects.html#./objects:h2)

Indeed, `symbol=?` isn't even as efficient as `eq?`

File: code/track.ss (hunk 4 above)
Comment (jitwit) [reply code/track.ss]: Sounds good. Snake case conversion stuff could go in `outils.ss` though

File: code/outils.ss (hunk 8 above)
Comment (guygastineau) [reply code/outils.ss]: Done

File: code/track.ss (hunk 4 above)
Comment (guygastineau) [reply code/track.ss]: Yeah, I kept it there at least for this PR.

File: code/outils.ss (hunk 8 above)
Comment (guygastineau) [reply code/outils.ss]: That is wild!

I wonder why it even exists then...

**Comment by guygastineau** [comment]
When this is ready to merge I should probably rebase it into just a couple commits.

**Comment by guygastineau** [comment]
What do you think about this PR now?

File: code/json.sls [hunk 9]
```diff
@@ -30,172 +30,233 @@
...
-		(not (pair? (caar x))))
-	   (write-ht x p))
-	  ((list? x) (write-array x p))
-	  (else (error "Invalid JSON object in json-write" x))))
+        (cond
+         ((or (string? x)
+              (number? x)) (write x p))
+         ((boolean? x) (display (if x "true" "false") p))
+         ((symbol? x)
+          (write (if (eq? x 'null) 'null (symbol->string x)) p))
```
Comment (jitwit) [inline code/json.sls]: I think `null` is represented as `'()`, so this check might be unnecessary. 
```
> (with-input-from-string "null" json-read)
()
```

File: code/track.ss [hunk 10]
```diff
@@ -40,7 +42,7 @@
 
 ;;; UUID
 
-;; wrapper to read uuid generated by configlet from scheme
+;; wrapper to read uuid generated by bonfiglet from scheme
```
Comment (jitwit) [inline code/track.ss]: ?

Comment (guygastineau) [reply code/track.ss]: 🤣 sorry for the typo.  I just rewrite it when it was moved.

I will fix it.

**Comment by jitwit** [comment]
I do wish that if you want to clean up formatting, tabs/spaces, and trailing whitespace that that happen in a separate pull request. Or even if you want to fix it that we agree on using the same emacs extensions. If I go back and edit something, it'll just revert to what I had before

File: code/json.sls (hunk 9 above)
Comment (guygastineau) [reply code/json.sls]: I think you are right.

This was already in the lib.

I am interested in refactoring this library (and adding better comments to the original author's work), but I think that a full refactor and documentation of his library might be better in a new PR for the future.

Is that fine with you if this change waits until then?  Without a test sweet designed to prove that refactoring doesn't break anything I am concerned there could be edge cases we break by refactoring his code further than the current latch goes.

Comment (jitwit) [reply code/json.sls]: Fair enough, yeah, sounds fine to keep!

**Comment by guygastineau** [comment]
Yes, I am sorry about that. I really don't know why the formatting comes in all messed up.

Does the indentation look normal on your machine?

**Comment by guygastineau** [comment]
Although, concerning trailing whitespace, I think trailing whitespace should not make it into the files at all.

**Comment by jitwit** [comment]
Yeah, the indentation looks normal! haha. It's weird that the github diffs show the tabs mangled, but in the source files they look normal. 

Also just to understand, what's your beef with trailing whitespace?

**Comment by guygastineau** [comment]
My mentors when I first started programming and getting involved with open source projects thought that it made the file sort of dirty/cluttered unnecessarily.

Since then I always have extensions that highlight trailing whitespace in red, so it is immediately apparent to me when present in files.

The way I see it, trailing whitespace is completely unnecessary, so there is no reason to let it linger.

Also, when I use a hotkey to go jump to the end of the line in insert mode I expect to jump to the position after the last significant character of the line.  Trailing whitespace ruins the efficiency of vimstyle composable editing commands.

I am not against a single trailing newline at the end of a sec code document though.

**Comment by guygastineau** [comment]
Interestingly, when I open files that have been edited by your Emacs they inner blocks are often not indented as far as their parent.

I will look into what the differences are, so we can move forward without this issue.

Basically it makes trying to read or edit the source code impossible in my Emacs until I reformat it.

**Comment by jitwit** [comment]
I don't think the conclusion "there's no reason to let it linger" follows from "whitespace is unnecessary".  

In terms of modes I have `aggressive-indent` and `paredit`. The issue could be that the `tab-width` variables are different? Mine seems to be set at 8

**Comment by guygastineau** [comment]
I'll look into those Emacs packages and the tab-width.
Just curious, if the trailing whitespace is unnecessary what would be the reason to keep it around?

**Comment by jitwit** [comment]
I was just commenting on the logical reasoning. I don't think there's a reason to keep it either! It's like vestigial organs, tail bones aren't necessary for humans, but that doesn't mean they should be gotten rid of

**Comment by guygastineau** [comment]
BTW it looks like I am using the electric indent package with Emacs.

Concerning the logic:

It depends how we relate `reason` to `necessity`.

If we are sologistic about it:

If there is a reason for something, then it is necessary.

There is a reason for cars, they are necessary.

That logic is certainly not correct.

I do think that we can say that something must have a reason in order to be necessary, but that also won't allow my assertion to stand up to Aristotelian proofs.

So, in short, you are right.  My assertion was not silogistically logical.

To rephrase what I meant.  'there is no practicality to leaving the trailing whitespace in a document'.

To extend the importance I have found cases wherein leaving the whitespace has negative consequences for efficiency of file editing.

So, while that first assertion was not probably logical, that does not detract from the fact that there do exist reasons to exclude (ie. not let it linger) whitespace.

Most vestigial limbs are dwarfed and over time the process of biological evolution is capable of removing them entirely.

I simply suggest that we take a proactive stance against an unnecessary inclusion that has tangible drawbacks.

**Comment by guygastineau** [comment]
So, it seems like the difference between electric and aggressive indentation shouldn't cause the formatting problem.  I am looking into the tab width right now.

I believe mine is set to `4` which in the spirit of a proper internet flame war is the correct tabwidth :rofl:

**Comment by guygastineau** [comment]
Well, I am confused about how that could have bearing, because the indentation rules for scheme-mode are much more complicated, and it is probably completely overriding any tabwidth variable that is set.

Do you use schem-mode with geiser and select an implementation of scheme while editing?

File: code/track.ss (hunk 10 above)
Comment (guygastineau) [reply code/track.ss]: Fixed

**Comment by guygastineau** [comment]
I am going to rebase this on top of `feature` in order to resolve the merge conflicts.

Then I will rebase this PR down to just a couple commits.

I guess we will need at least a final pass after that to make sure that the conflict resolution turned out alright.

**Comment by guygastineau** [comment]
I just used interactive rebase to whittle this one down while also fixing conflicts all in one go.

**Comment by guygastineau** [comment]
I am still stumped by what is happening with formatting.  GitHub seems to understand my formatting, but somehow it thinks that your tabs are not wide enough it seems.  Although it is also clear in the diffs that both of our emacs are using variable width tabs to deal with the very complicated indentation rules of `Scheme`.

I mean I like the indentation rules. I think they make the code more friendly to humans (and generally more aesthetically pleasing).  I am just still really confused about why our editors aren't doing it the same.  At the same time, some files don't have that happening.  Like the `config.ss` file isn't having any formatting issues.  The big part of the diff there is from removing the trailing whitespace.

**Comment by guygastineau** [comment]
What's your take on this now @jitwit ?

File: code/config.ss [hunk 11]
```diff
@@ -2,67 +2,79 @@
...
+      (difficulty . 3)
+      (topics list))
+
+     ((slug . grains)
+      (uuid . "397d9cfc-43f9-4c57-bc55-5c94ae8f532e")
+      (core . #t)
+      (unlocked-by)
+      (difficulty . 3)
+      (topics math))
+
```
Comment (guygastineau) [inline code/config.ss]: Sorry. Somehow I messed this up in the conflict resolution. I will fix it.

Comment (guygastineau) [reply code/config.ss]: Last commit fixed it.

**Comment by guygastineau** [comment]
The `config.json` was out of sync with the `config.ss` on feature.

Now everything in this PR is good to go.

File: config.json [hunk 12]
```diff
@@ -141,7 +141,9 @@
       "slug": "nucleotide-count",
       "uuid": "d72cf2c3-9f30-4df1-a69b-04c638bb2426",
       "core": false,
-      "unlocked_by": null,
+      "unlocked_by": [
```
Comment (guygastineau) [inline config.json]: Actually, these unlocked by sections are supposed to be strings, not arrays.

**Comment by guygastineau** [comment]
Okay, so my last commit changed the `unlocked-by` field of the `config.ss` so that it is a `key/value` pair instead of a list.  This is more appropriate as there can only be one value.  Also, this way it gets rendered to the correct json automatically.

Are there any other parts of the tooling you have made that will break from this?

File: code/config.ss [hunk 13]
```diff
@@ -78,63 +76,63 @@
...
       (core . #f)
       (unlocked-by rna-transcription)
       (difficulty . 3)
       (topics string list))
-     
+
      ((slug . word-count)
       (uuid . "9460b65d-dc80-4a95-8782-b395d2cc979e")
       (core . #f)
       (unlocked-by nucleotide-count)
```
Comment (guygastineau) [inline code/config.ss]: Travis is breaking because of this line.

It is a good reminder that all changes to `config.ss` should have `config.json` regenerated, so that `configlet` can check for incompatibilities.

What do you think is a good core exercise to use for unlocking `word-count`? It must be core.

Comment (guygastineau) [reply code/config.ss]: I think that `anagram` would be a good choice for unlocking `word-count` given the current list of core exercises.

I will go ahead and change that so this is passing.  let me know what you think about it.

Comment (guygastineau) [reply code/config.ss]: Made the change

**Comment by guygastineau** [comment]
So, I would have preferred not having to change the dependency for `word-count` or having to change the `unlocked-by` fields to pairs in this PR, but I was already really far into this, so I am just trying to get this ready to merge, so I don't have to do any more damn merge conflicts on it.

I think everything is actually ready to go now.  I am just hoping for another set of eyes on it one last time, before it gets merged.

File: config.json (hunk 12 above)
Comment (jitwit) [reply config.json]: really? configlet hasn't been complaining

File: code/config.ss (hunk 13 above)
Comment (jitwit) [reply code/config.ss]: I always make before pushing

File: config.json (hunk 12 above)
Comment (jitwit) [reply config.json]: I see you're right. let's not make this change here lol. I'm adding problems in other pull requests and I think it could lead to conflicts

File: code/config.ss (hunk 13 above)
Comment (guygastineau) [reply code/config.ss]: I'm not sure why the exercise dependencies weren't being added to the `config.json` then.

Oh well. What's your take on this PR now?

I think it is ready (even if it isn't perfect).

It will probably be easier to add future improvements once this is no longer pending.

Comment (jitwit) [reply code/config.ss]: The biggest requirements I guess are no conflicts and the output matches `confliglet`'s formatting

Comment (jitwit) [reply code/config.ss]: Ah they weren't being added because of scheme-case. Your pull will change that. (Currently setting `unlocked-by` to `unlocked_by` works, too)

Comment (guygastineau) [reply code/config.ss]: Cool. So are you ready for this to get merged?

**Comment by guygastineau** [comment]
Thanks.
